src/tree/builder/TreeAdapter.js

```diff
@@ -0,0 +1,449 @@
+	 *
+	 * @returns {void}
+	 */
+	appendChild( parent, child ) {
+		// Do not do anything with irrelevant content.
+		if ( parent instanceof StructuredIrrelevant ) {
+			return;
+		}
+
+		if ( child instanceof StructuredIrrelevant &&
+			( TreeAdapter._isLeafNode( child ) || parent instanceof FormattingElement ) ) {
+			// Add StructuredIrrelevant element as formatting to the parent node.
+			const element = new FormattingElement( child.tagName, {} );
+			element.location = child.location;
+			this._appendFormattingElement( parent, element );
+			return;
+		}
+
+		if ( child instanceof FormattingElement ) {
+			this._appendFormattingElement( parent, child );
+			return;
+		}
+
+		// Just add nodes to parent's children if not formatting.
+		child.parent = parent;
+		parent.children.push( child );
+	}
+
+	/**
+	 * Appends the formatting element to the tree.
+	 *
+	 * @param {Node} parent                          The (current) parent of the formatting element.
+	 * @param {FormattingElement} formattingElement  The formatting element to add to the tree.
+	 *
+	 * @returns {void}
+	 *
+	 * @private
+	 */
+	_appendFormattingElement( parent, formattingElement ) {
+		formattingElement.parent = parent;
+		if ( parent instanceof StructuredNode ) {
+			// Wrap it in a paragraph, add it as formatting.
+			TreeAdapter._addOrphanedFormattingElement( formattingElement, parent );
+		} else {
+			/*
+			 Formatting elements can be nested, we want to add it to
+			 the most recent ancestor which is either a heading or paragraph.
+			 */
+			const ancestor = TreeAdapter._findAncestor( formattingElement, TreeAdapter._isLeafNode );
+			if ( TreeAdapter._isLeafNode( ancestor ) ) {
+				// Add formatting element as formatting to the found paragraph or heading ancestor.
+				formattingElement.parent = parent;
+				ancestor.textContainer.formatting.push( formattingElement );
+			} else {
+				// Wrap formatting element in paragraph, add it to the tree.
+				TreeAdapter._addOrphanedFormattingElement( formattingElement, parent );
+			}
+		}
+	}
+
+	/**
+	 * Wraps a formatting element in a paragraph and adds the resulting paragraph to the given parent.
+	 *
+	 * @param {FormattingElement} formattingElement  The formatting element to wrap in a paragraph and add to the tree.
+	 * @param {Node} parent                          The parent element to add the new paragraph to.
+	 *
+	 * @returns {void}
+	 *
+	 * @private
+	 */
+	static _addOrphanedFormattingElement( formattingElement, parent ) {
+		const paragraph = new Paragraph();
+		paragraph.location = formattingElement.location;
+
+		paragraph.textContainer.formatting.push( formattingElement );
+
+		paragraph.parent = parent;
+		formattingElement.parent = paragraph;
+		parent.children.push( paragraph );
+	}
+
+	/**
+	 * Detaches a node from its parent.
+	 *
+	 * @param {Node} node	The node to detach from its parent.
+	 *
+	 * @returns {void}
+	 */
+	detachNode( node ) {
+		if ( node.parent ) {
```

Comment:
It seems the else statement of this if is never reached in the coverage files.

Reply:
I added two tests in `TreeAdapterSpec.js` that tests the `detachNode` functionality (both in the case of having a parent and not having one).